feat(signals): Enrich transcription with metadata

sortafreel opened the pull request:
## Problem

- Can't track the Golden path at scale with free text, need chains of events.
- Summaries (and segments/points within) aren't searchable without embeddings.
- Including autocapture events in video-first summaries prompts heavily bloated the cost, not scalable.

## Changes

- As I significantly reduced the problems with unreliable clicks/timestamps, we can rely way more on DB data, instead of LLM.
- I updated LLM video-first prompts to return labels of interactable elements/blocks/pages together with the transcription.
- I heavily filtered autocapture events and am now able to attach them properly to interactions within video segments, without the need to feed DB data into the LLM.

Now, in addition to the regular transcription (left, which can be used for further summaries), we are also getting autocapture-like events to store in DB, to mark clicks/hovers/navigation/loading/input/etc. (right). And all of it is crazy cheap and scalable.

Events are structured and can be linked to ts/URLs/event ids/etc. When I'm back, I'll start storing them and use them for non-embedding search and as a timeseries to calculate Golden paths and deviations.

**Bonus:** Because event chains are built explicitly on the labels/page titles/etc. of the actual website, we also get a reliable taxonomy for feature detection.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changelog: (features only) Is this feature complete?

<!-- Yes if this is okay to go in the changelog. No if it's still hidden behind a feature flag, or part of a feature that's not complete yet, etc.  -->
<!-- Removing this section does not mean the changelog bot won't pick it up, because *some people* like to not use the template, so we can't rely on it existing. -->

Twixes committed:
feat(signals): Video-first session summarization

Twixes committed:
feat(signals): Video-first session summarization

Twixes committed:
Add segment consolidation step

Twixes committed:
Add segment consolidation step

Twixes committed:
Remove test code

Twixes committed:
Remove test code

Twixes committed:
Split up activities into one per file

Twixes committed:
Split up activities into one per file

Twixes committed:
Add test command

Twixes committed:
Add test command

Twixes committed:
Use shared util for duration extraction

Twixes committed:
Use shared util for duration extraction

Twixes committed:
Rename files for etra clarity

Twixes committed:
Rename files for etra clarity

Twixes committed:
Update nodes.py

Twixes committed:
Update nodes.py

Twixes committed:
Skip sessions too short and skip first 2s

Twixes committed:
Skip sessions too short and skip first 2s

Twixes committed:
Refactor prompts and improve tracing

Twixes committed:
Refactor prompts and improve tracing

Twixes committed:
Align final output with original schema

Twixes committed:
Align final output with original schema

Twixes committed:
DRY the code

Twixes committed:
DRY the code

Twixes committed:
Address Greptile comments

Twixes committed:
Address Greptile comments

Twixes committed:
Update test_summarize_session_group.py

Twixes committed:
Update test_summarize_session_group.py

Twixes committed:
Update a2_upload_video_to_gemini.py

Twixes committed:
Update a2_upload_video_to_gemini.py

Twixes committed:
Update summarize_session.py

Twixes committed:
Update summarize_session.py

Twixes committed:
Update test_session_summaries.py

Twixes committed:
Update test_session_summaries.py

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
fix: Don't block the loop with time.

sortafreel committed:
fix: Don't block the loop with time.

sortafreel committed:
fix: Increase timeout for video export, as we use 1x.

sortafreel committed:
fix: Increase timeout for video export, as we use 1x.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
fix: Log exceptions with tag.

sortafreel committed:
fix: Log exceptions with tag.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
feat: Highlight clicks.

sortafreel committed:
feat: Provide clicks highlight through URL params.

sortafreel committed:
feat: Highlight clicks for session moments.

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'video-based-session-summaries' of github.com:PostHog/posthog into video-based-session-summaries

sortafreel committed:
chore: Test comment

sortafreel committed:
chore: Remove test comment.

sortafreel committed:
feat: Add metadata to segments.

sortafreel committed:
chore: Test change.

sortafreel committed:
feat: Add overlay to display more context in the exported recording.

sortafreel committed:
feat: Use mono font.

sortafreel committed:
fix: Rename.

sortafreel committed:
feat: Enable LLM meta if provided through URL params.

sortafreel committed:
feat: Enable LLM-highlighted clicks as a default animation.

sortafreel committed:
Merge branch 'signals/highlight-clicks' into 12-26-feat_signals_enrich_transcription_with_metadata

sortafreel committed:
Merge branch 'signals/show-url' into 12-26-feat_signals_enrich_transcription_with_metadata

sortafreel committed:
feat: Update prompt to care about black line.

sortafreel committed:
chore: Add comment.

sortafreel committed:
fix: Change timestamp naming to don't conflict with Gemini timestamps.

sortafreel committed:
feat: Calculate offset and combine interactions.

sortafreel committed:
feat: Attach clicks and contexts to the segments.

sortafreel committed:
feat: Increase the chunk size.

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'video-based-session-summaries' of github.com:PostHog/posthog into video-based-session-summaries

sortafreel committed:
chore: Remove excessive comment.

Twixes committed:
Remove unused `asset_id`

Twixes committed:
Add flush

Twixes committed:
Handle None duration

Twixes committed:
Addressing type notes

Twixes committed:
Add upload timeout

Twixes committed:
Add comment to each activity file

Twixes committed:
Reduce nesting

Twixes committed:
Use DEFAULT_VIDEO_UNDERSTANDING_MODEL

Twixes committed:
Make skipping logic for too-short videos more robust

Twixes committed:
Use AsyncClient

This requires a `posthoganalytics.ai` upgrade

Twixes committed:
Consolidate empty `raw_segments` handling

Twixes committed:
Remove needless _convert_video_segments_to_session_summary_fallback

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Add VideoSegmentOutcome dataclass

Twixes committed:
Clarify _format_timestamp_as_mm_ss

Twixes committed:
Be Pydantic

Twixes committed:
Speed up video playback a bit

Twixes committed:
Validate end_time vs. start_time

Twixes committed:
Restore try:catch:

Twixes committed:
Tune logging levels

Twixes committed:
Improve segment error handling

Twixes committed:
Optimize team_name

Twixes committed:
Fix stray inline import

Twixes committed:
Revert "Optimize team_name"

This reverts commit 36d4ebc198294ffe2bb155e2dcb20323f74e146a.

Twixes committed:
Address AI review points

Twixes committed:
Make mypy happy

Twixes committed:
Improve "Static"

Twixes committed:
Make example a bit more varied

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Gilbert09 commented on posthog/hogql/database/s3_table.py:
```diff
@@ -21,6 +23,10 @@ def build_function_call(
 ) -> str:
+    if access_key is None and access_secret is None and (settings.DEBUG or settings.TEST):
+        access_key = settings.AIRBYTE_BUCKET_KEY
+        access_secret = settings.AIRBYTE_BUCKET_SECRET
```
It looks like you've managed to pull in a bunch of changes from `master` in here - think a rebase is required before a review

Twixes replied:
Yup, fixed, my bad!

Twixes committed:
Merge branch 'video-based-session-summaries' into 12-26-feat_signals_enrich_transcription_with_metadata

Twixes committed:
Remove unused file

Twixes committed:
Fix type issues

Twixes committed:
Make segment consolidation preserve more detail

Twixes committed:
Address inaccurate log

Twixes committed:
Make mypy happy

Twixes committed:
Fix todo

Twixes committed:
Remove max_output_tokens

This was causing cut off tokens. The output is now larger internally due to us using reasoning.

Twixes commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -224,95 +194,354 @@ async def analyze_video_segment_activity(
+                        signals_type="session-summaries",
+                    )
+                    element_type = VideoSegmentTypesEnum.CUSTOM.value
+                elements.append(VideoSegmentElement(element_type=element_type, element_value=value.strip('"')))
```
The format is kind of odd with these additions
```
- **00:02 - 00:02**: The user clicks on the "View pricing" button on the Hedgebox landing page.
  `elements: button="View pricing"`
  `interaction: click`
  `rec_t: 3`
```
Should we just use good ol YAML?

Twixes commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -194,123 +166,364 @@ async def analyze_video_segment_activity(
+                session_start_time=session_start_time,
+            )
+            if not events_in_range:
+                return None
```
+1 to the bot

Twixes reviewed (approved):
Looks good to me, this is an interesting approach! Before we merge: one issue highlighted by Greptile.

A few TODOs I see, but makes sense to leave them out here.

I did push a few commits to fix small things, hope they're all likable.
